devito/operations/interpolators.py

```diff
@@ -388,3 +389,520 @@ def callback():
+            idx_subs, temps = self._interpolation_indices(variables, offset,
+                                                          field_offset=field_offset)
+
+            rhs = Symbol(name='sum', dtype=self.sfunction.dtype)
+            summands = [Eq(rhs, 0., implicit_dims=self.sfunction.dimensions)]
+
+            dim_pos = self.sfunction.grid.dimensions
+
+            # Checks if the data is 3D or 2D
+            if len(dim_pos) == 3:
+                eqs = [np.sum(v) for v in self._tricubic_equations(_expr,
+                                                                   idx_subs,
+                                                                   dim_pos=dim_pos)]
+            else:
+                eqs = [np.sum(v) for v in self._bicubic_equations(_expr,
+                                                                  idx_subs,
+                                                                  dim_pos=dim_pos)]
+
+            # Creates Sympy equation using the values in eq as
+            # the right side of the equation
+            summands.extend([Inc(rhs, v, implicit_dims=self.sfunction.dimensions)
+                             for v in eqs])
+
+            # Write/Incr `self`
+            lhs = self.sfunction.subs(self_subs)
+            last = [Inc(lhs, rhs)] if increment else [Eq(lhs, rhs)]
+
+            return temps + summands + last
+
+        return Interpolation(expr, offset, increment, self_subs, self, callback)
+
+    def inject(self, field, expr, offset=0):
+        """
+        Generate equations injecting an arbitrary expression into a field.
+        Parameters
+        ----------
+        field : Function
+            Input field into which the injection is performed.
+        expr : expr-like
+            Injected expression.
+        offset : int, optional
+            Additional offset from the boundary.
+        """
+        def callback():
+            # Derivatives must be evaluated before the introduction of indirect accesses
+            try:
+                _expr = expr.evaluate
+            except AttributeError:
+                # E.g., a generic SymPy expression or a number
+                _expr = expr
+
+            variables = list(retrieve_function_carriers(_expr)) + [field]
+
+            # Need to get origin of the field in case it is staggered
+            field_offset = field.origin
+
+            # List of indirection indices for all adjacent grid points
+            idx_subs, temps = self._interpolation_indices(variables, offset,
+                                                          field_offset=field_offset)
+
+            dim_pos = self.sfunction.grid.dimensions
+            if len(dim_pos) == 3:
+                eqs = [v for v in self._tricubic_equations(_expr,
+                                                           idx_subs,
+                                                           dim_pos=dim_pos)]
+            else:
```

Comment:
Can we do this with inheritance? Or in a way that generalises beyond 2/3 dimensions?